distribution/pull_v2.go

```diff
@@ -858,6 +859,66 @@ func fixManifestLayers(m *schema1.Manifest) error {
 	return nil
 }
 
-func createDownloadFile() (*os.File, error) {
-	return ioutil.TempFile("", "GetImageBlob")
+var v2LayerDataFilename = "v2data"
+
+func resumeOrCreateDownloadFile(progressOutput progress.Output, cachePath string, dgst digest.Digest, id string) (int64, *os.File, digest.Verifier, error) {
+	dataPath := filepath.Join(cachePath, v2LayerDataFilename)
+	_, err := os.Stat(dataPath)
+	if err == nil {
+		return resumeDownloadFile(progressOutput, cachePath, dgst, id)
+	}
+	if os.IsNotExist(err) {
+		tmpFile, err := createDownloadFile(cachePath)
+		if err != nil {
+			return 0, nil, nil, err
+		}
+		return 0, tmpFile, nil, nil
+	}
+	return 0, nil, nil, err
+}
+
+func resumeDownloadFile(progressOutput progress.Output, cachePath string, dgst digest.Digest, id string) (int64, *os.File, digest.Verifier, error) {
+	dataPath := filepath.Join(cachePath, v2LayerDataFilename)
+	verifier, err := digest.NewDigestVerifier(dgst)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	// the last parameter here is ignored because this mode never creates a file
+	file, err := os.OpenFile(dataPath, os.O_RDWR, 0)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+	info, err := file.Stat()
+	if err != nil {
+		return 0, nil, nil, err
+	}
+	size := info.Size()
+
+	// TODO(vikstrous) make a mechanism for storing the partial digest and offset atomically
+	// and resuming using that instead of re-hashing everything up to the current
+	// offset
```

Comment:
Sounds like if we wanted to do this, we would have to switch to these resumable hashes in distribution. I wouldn't want to do this in docker without doing it in distribution first. Also, it looks like re-extracting layers will be way slower than re-hashing, so we should solve that first if we want to speed things up.